typescript/alert-center/src/utilities.ts

```diff
@@ -0,0 +1,165 @@
+import { ArgOptional, Transaction, TransactionContext} from '@dbos-inc/dbos-sdk';
+import { Knex } from 'knex';
+
+type KnexTransactionContext = TransactionContext<Knex>;
+
+export enum AlertStatus {
+  ACTIVE   = 0,
+  ASSIGNED = 1,
+  RESOLVED = 2
+}
+
+export interface Employee {
+  employee_name: string;
+  alert_id: number | null;
+  expiration: Date | null;
+  timeLeft?: number;
+}
+
+export interface AlertEmployee {
+  alert_id: number;
+  alert_status: AlertStatus;
+  message: string;
+  employee_name: string | null;
+}
+
+export interface AlertWithMessage {
+  alert_id: number;
+  alert_status: AlertStatus;
+  last_update_time: Date;
+  message: string;
+}
+
+export class RespondUtilities {
+  @Transaction({readOnly: true})
+  static async popDashboard(ctx: KnexTransactionContext) {
+    const alerts = await ctx.client<AlertEmployee>('alert_employee').select().orderBy(['alert_id']);
+    const employees = await ctx.client<Employee>('employee').select().orderBy(['employee_name']);
+    for (const a of alerts) {
+      if (a.alert_status === AlertStatus.ACTIVE && a.employee_name) {
+        a.alert_status = AlertStatus.ASSIGNED;
+      }
+    }
+    for (const p of employees) {
+      if (p.expiration) {
+        p.timeLeft = Math.round((p.expiration.getTime() - new Date().getTime())/1000);
+      }
+    }
+    return {alerts, employees};
+  }  
+
+  @Transaction()
+  static async cleanStaff(ctx: KnexTransactionContext) {
+    await ctx.client<Employee>('employee').whereNull('alert_id').delete();
+  }
+
+  @Transaction()
+  static async cleanAlerts(ctx: KnexTransactionContext) {
+    await ctx.client<AlertEmployee>('alert_employee').where({alert_status: AlertStatus.RESOLVED}).delete();
+  }
+
+  @Transaction()
+  static async getMaxId(ctx: KnexTransactionContext) {
+    const result = await ctx.client<AlertEmployee>('alert_employee').max('alert_id', { as: 'mid' }).first();
+    if (result ) {
+      return result.mid;
+    }
+    return -1;
+  }
+
+  @Transaction()
+  static async addAlert(ctx: KnexTransactionContext, message: AlertWithMessage) {
+    await ctx.client<AlertEmployee>('alert_employee').insert({
+      alert_id: message.alert_id,
+      alert_status: message.alert_status,
+      message: message.message,
+      employee_name: null,
+    }).onConflict(['alert_id']).ignore();
+  }
+
+  @Transaction()
+  static async getUserAssignment(ctx: KnexTransactionContext, employee_name: string, expiration: Date, @ArgOptional more_time: boolean | undefined) {
+    let employees = await ctx.client<Employee>('employee').where({employee_name}).select();
+    let newAssignment = false;
+    if (!employees.length) {
+      await ctx.client<Employee>('employee').insert({employee_name, alert_id: null, expiration: null});
```

Comment:
`employee_name` is the primary key of that table. So inserting the same name multiple time would fail. We can add a on conflict `employee_name` "ignore" here.

Reply:
But this is in a transaction where we just queried for employee name (see above). So this kind of error should never happen?

Reply:
You're right. Our default isolation level is serializable so this would cause a serialization error and be retried. Then during retry, the transaction will see the inserted data.